test/cpp/jit/test_gpu.cpp

```diff
@@ -10039,6 +10039,318 @@ TEST(NVFuserTest, FusionGetComputeAtRelPos_CUDA) {
+
+  // compare with reference output
+  auto cat_tensor = at::cat({init_in, in0}, 1);
+  TORCH_CHECK(cat_tensor.var({1}, false).allclose(out_var));
+  TORCH_CHECK(
+      cat_tensor.mean({1}).allclose(out_avg, /*rtol*/ 1e-5, /*atol*/ 1e-6));
+}
+
+TEST(NVFuserTest, blockWelfordNoInit) {
+  FusionExecutor fe;
+  int x = 7, y = 8, z = 9;
+
+  // need support IValue for integer input as initial count
+  std::string kernel = R"(
+__global__ void kernel1(
+    Tensor<float,3> inp,
+    Tensor<float,1> out_var,
+    Tensor<float,1> out_avg
+){
+    //actual generated kernel will use dynamic shared mem,
+    // here is just for prototype
+    __shared__ float mem_M2[1024];
+    __shared__ float mem_avg[1024];
+    float in=inp[threadIdx.x*inp.stride[0]+
+                        threadIdx.y*inp.stride[1]+
+                        threadIdx.z*inp.stride[2]];
+    float tmp_M2;
+    float tmp_avg;
+    int tmp_N;
+    blockWelford<false,true,true>(
+        tmp_M2,
+        tmp_avg,
+        tmp_N,
+        0.f,
+        in,
+        1,
+        threadIdx,
+        blockDim,
+        (float*)mem_M2,
+        (float*)mem_avg,
+        (bool)(threadIdx.x<inp.size[0]),
+        0.f);
+    __syncthreads();
+    if(threadIdx.x<out_var.size[0] && threadIdx.y==0 && threadIdx.z==0){
+        out_var[threadIdx.x*out_var.stride[0]]=tmp_M2/(tmp_N);
+        out_avg[threadIdx.x*out_var.stride[0]]=tmp_avg;
+    }
+}
+    )";
+  fe.compileRTC(kernel, "CudaCodeGen::kernel1");
+  LaunchParams lp(
+      1, // gdimx
+      1, // gdimy
+      1, // gdimz
+      x, // bdimx
+      y, // bdimy
+      z // bdimz
+  );
+  lp.setSmem(0);
+  const auto options =
+      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  const std::vector<int64_t> tensor_dims = {x, y, z};
+  auto in0 = at::randn(tensor_dims, options);
+  auto out_var = at::empty({x}, options);
+  auto out_avg = at::empty({x}, options);
+  fe.runRTC(lp, {in0, out_var, out_avg});
+
+  TORCH_CHECK(in0.var({1, 2}, false).allclose(out_var));
+  TORCH_CHECK(in0.mean({1, 2}).allclose(out_avg, /*rtol*/ 1e-5, /*atol*/ 1e-6));
+}
+
+TEST(NVFuserTest, gridWelfordNoInit) {
+  FusionExecutor fe;
+  int x = 128, y = 64, z = 128;
+
+  std::string kernel = R"(
+__global__ void kernel1(
+    Tensor<float,3> inp,
+    Tensor<float,1> out_var,
+    Tensor<float,1> out_avg,
+    Tensor<float,1> work_buf_M2,
+    Tensor<float,1> work_buf_avg,
+    Tensor<int64_t,1> sync_flag
+){
+    __shared__ float shared_buf_M2[1024];
+    __shared__ float shared_buf_avg[1024];
+    float tmp_M2;
+    float tmp_avg;
+    int tmp_N;
+    float in = inp[ blockIdx.x  * inp.stride[0]+
+                    blockIdx.y  * inp.stride[1]+
+                    threadIdx.x * inp.stride[2]];
+    bool T_pred;
+    T_pred=welford::gridWelford<
+        true,true,false,
+        true,false,false
+    >(
+        tmp_M2,
+        tmp_avg,
+        tmp_N,
+        0.f,
+        in,
+        1, 
```

Comment:
I think passing the `inp_N = 1` require all threads to have the same pixels to cover. Could be an awkward requirement for evil shapes.

Reply:
No, I was wrong.
I think it only impacts scheduling? Are we going to support a serial welford within a thread? That's the only place where it *could* cause inconsistent count on threads. So this will only be a limiting factor for scheduling.

I think for now it's OK to leave it as-is. If flexibility for scheduling is needed later we can refactor this.

Reply:
We can ignore that value like we do inputs based on the provided predicate for block/grid welford reductions. Single thread welford would have the predicate around it. I don't think it should be an issue, though not 100% sure.

Reply:
Actually on a second thought. I temporarily decided to fall back to shared mem approach for keeping track of N, to avoid limiting scheduling options at too early stage. But still in this case shared mem will be filled with very repetitive data. Will revisit the compression on N part after scheduling on Welford op works.